### ISSUE TITLE
[GraphBolt] ItemSampler hetero example uses too much CPU

### ISSUE THREAD

**Issue by mfbalin** [dmlc/dgl#7315]
## 🔨Work Item

**IMPORTANT:**
* This template is only for dev team to track project progress. For feature request or bug report, please use the corresponding issue templates.
* DO NOT create a new work item if the purpose is to fix an existing issue or feature request. We will directly use the issue in the project tracker.

Project tracker: https://github.com/orgs/dmlc/projects/2

## Description

When running the hetero graphbolt example in the pureGPU mode, the CPU utilization is very high. My suspected cause is the ItemSampler. Because we have a prefetcher right after the ItemSampler, it can run at the same time as the rest of the sampling pipeline running on the GPU, the profile seems to verify this claim.

![image](https://github.com/dmlc/dgl/assets/16190101/50f04dae-8835-44f7-a96b-bf5402294336)

**Comment by Rhett-Ying** [comment]
As @mfbalin mentioned, specific logic for `ItemsetDict` could be the culprit.

**Comment by mfbalin** [comment]
dgl/heterograph.py:6407 make_canonical_edges uses numpy for some ops.

**Comment by Rhett-Ying** [comment]
https://github.com/dmlc/dgl/blob/41a38486a5ed9298093d9f0bc415751269c7d577/python/dgl/convert.py#L583

**Comment by Rhett-Ying** [comment]
@peizhou001 
can `dgl.create_block()` purely run on GPU or it can only run purely on CPU instead? I remember you looked into it previously.

**Comment by mfbalin** [comment]
Update, CPU usage high even for the homo examples. Some recent change might have caused us to utilize the CPU even in the pureGPU mode. @frozenbugs do you think it could be the logic to move MiniBatch to device?

**Comment by mfbalin** [comment]
Or could it possibly be one of my recent changes, such as #7312?

**Comment by mfbalin** [comment]
`git checkout 78df81015a9a6cdaa4843167b1d000f4ca377ca9`

This commit does not have the issue. Somewhere between current master and the reported commit above, there was a change that cause CPU util on the GPU code path.

**Comment by Rhett-Ying** [comment]
> `git checkout 78df81015a9a6cdaa4843167b1d000f4ca377ca9`
> 
> This commit does not have the issue. Somewhere between current master and the reported commit above, there was a change that cause CPU util on the GPU code path.

@yxy235 could you help look into it? Reproduce and confirm it?

**Comment by mfbalin** [comment]
Easiest way to test is to run `python examples/sampling/graphbolt/pyg/node_classification_advanced.py --torch-compile --mode=cuda-cuda-cuda`. There is upto %30 regression.

**Comment by mfbalin** [comment]
Transfered attr list:
`['blocks', 'compacted_negative_dsts', 'compacted_negative_srcs', 'compacted_node_pairs', 'compacted_seeds', 'edge_features', 'indexes', 'input_nodes', 'labels', 'negative_dsts', 'negative_node_pairs', 'negative_srcs', 'node_features', 'node_pairs', 'node_pairs_with_labels', 'positive_node_pairs', 'sampled_subgraphs', 'seed_nodes', 'seeds']`

**Comment by yxy235** [comment]
> Transfered attr list:
> 
> ```
> ['blocks', 'compacted_negative_dsts', 'compacted_negative_srcs', 'compacted_node_pairs', 'compacted_seeds', 'edge_features', 'indexes', 'input_nodes', 'labels', 'negative_dsts', 'negative_node_pairs', 'negative_srcs', 'node_features', 'node_pairs', 'node_pairs_with_labels', 'positive_node_pairs', 'sampled_subgraphs', 'seed_nodes', 'seeds']
> 
> compacted_negative_dsts
> compacted_negative_srcs
> compacted_node_pairs
> compacted_seeds
> edge_features
> indexes
> input_nodes
> labels
> negative_dsts
> negative_srcs
> node_features
> node_pairs
> sampled_subgraphs
> seed_nodes
> seeds
> ```
> 
> Actually transfered by calling `.to`:
> 
> ```
> input_nodes
> labels
> seeds
> ```

I see. Do you think we need a check when call Minibatch.to()?

**Comment by mfbalin** [comment]
I figured it out. When we filter which attributes to transfer, we end up calling blocks property. Making a quick patch now.

**Comment by mfbalin** [comment]
CPU usage still higher than 100% though, so I am not sure if I resolved the whole issue.

**Comment by mfbalin** [comment]
Even with #7330, we need to investigate where the high CPU usage comes from.

**Comment by mfbalin** [comment]
hetero example CPU usage is still 4000%

**Comment by mfbalin** [comment]
Here, we can find see the last iterations of training datalaoder. Since we have a prefetcher thread with a buffer size 2, the last 2 iterations don't have excessive CPU utilization as the computation for the last 2 iterations has already finished. This indicates that the high CPU utilization is due to the ItemSampler.

```
        # (4) Cut datapipe at CopyTo and wrap with prefetcher. This enables the
        # data pipeline up to the CopyTo operation to run in a separate thread.
        datapipe_graph = _find_and_wrap_parent(
            datapipe_graph,
            CopyTo,
            dp.iter.Prefetcher,
            buffer_size=2,
        )
```

![image](https://github.com/dmlc/dgl/assets/16190101/0b28834b-2a32-4edf-9e4b-bdbd1fd7a381)

**Comment by mfbalin** [comment]
Users with multiple GPUs may not be able to utilize the GPUs effectively due to the CPU bottleneck.

**Comment by mfbalin** [comment]
`python examples/graphbolt/pyg/labor python node_classification.py --dataset=yelp --dropout=0 --mode=cuda-cuda-cuda`
CPU usage on this example is too high too, and this is the homo case. @Rhett-Ying
Becomes the bottleneck, faster CPU results into faster performance even if the GPU is slower.

**Comment by Rhett-Ying** [comment]
@mfbalin The culprit of high CPU usage is ItemSampler?
